From 457e03808f47c5939f30204658e20c0659970aee Mon Sep 17 00:00:00 2001
From: Jeremiah Orians <jeremiah@pdp10.guru>
Date: Tue, 27 Oct 2020 12:57:37 -0400
Subject: [PATCH 16/17] Fix broken test

---
 test/test3/hello.sh | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/test/test3/hello.sh b/test/test3/hello.sh
index 3d4196c..d5caab0 100755
--- a/test/test3/hello.sh
+++ b/test/test3/hello.sh
@@ -14,7 +14,7 @@
 ##
 ## You should have received a copy of the GNU General Public License
 ## along with stage0.  If not, see <http://www.gnu.org/licenses/>.
-set -eux
+set -ex
 ./bin/M1 -f test/test3/defs -f test/test3/lisp.s --BigEndian --architecture knight-native -o test/test3/hold
 ./bin/hex2 -f test/test3/hold --BigEndian --architecture knight-native --BaseAddress 0 -o test/results/test3-binary
 
-- 
2.20.1

